### ISSUE TITLE
Rules are executed as localrules although not declared as such

### ISSUE THREAD

**Issue by lhennig** [snakemake/snakemake#2048]
I have a snakemake workflow that is supposed to run a particle physics simulation pipeline repeatedly until a user-specified number of triggered events is reached. I implemented this by splitting the task in two parts: the simulation pipeline is summarized in a group job called `group0`, and the "calling the simulation pipeline repeatedly until..." part is implemented as a checkpoint that checks whether enough triggered events were produced. If yes, the workflow is finished. If no, the simulation pipeline runs again. The checkpoint is declared as a local rule, the simulation pipeline is not.

When I start my snakemake workflow from the frontend node of our cluster (not the best practice, but we are allowed to do so), everything runs fine in the beginning. A lot of `group0` jobs are sent to the cluster, after some time the checkpoints are evaluated again (locally on the frontend node), new `group0` jobs are sent out again. But after a while, I realized that some rules that are in the `group0` are run locally on the frontend node. I saw this by using the `htop` command, and I attached a screenshot of it below. There, you can see that some km3sim jobs were executed, although the rule km3sim is not declared as a local rule.

I attached my Snakefile and some python scripts used by the workflow. I also attached the output of the `htop` command where you can see that some km3sim jobs were run locally although they were never declared as localrules, and I also attached the output of snakemake at a time where it scheduled a km3sim and also a h5extractf job to run locally. This is, unfortunately, not a very good minimal example so far. 

![htop](https://user-images.githubusercontent.com/95235950/210805233-fb868ebf-0705-41af-8d17-5ff4ce9a09b0.png)

![snakemake_output](https://user-images.githubusercontent.com/95235950/210805256-aeebab4b-f04c-42a8-8d8f-6cbae1474707.png)

I uploaded the following files with a .txt extension such that I can upload them to github. Otherwise it complains about an unknown file format. But the names of the files have the correct .smk extension:
[Snakefile](https://github.com/snakemake/snakemake/files/10352958/Snakefile.txt)
[utility_functions.smk](https://github.com/snakemake/snakemake/files/10352961/utility_functions.txt)
[input_functions.smk](https://github.com/snakemake/snakemake/files/10352976/input_functions.txt)

[functions_for_snakefile.py](https://github.com/snakemake/snakemake/files/10352981/functions_for_snakefile.txt)
[create_checkpoint_summary.py](https://github.com/snakemake/snakemake/files/10353039/create_checkpoint_summary.txt)

**Comment by lhennig** [comment]
One additional note: this problem occurs when I use the native slurm support as well as when I use the generic cluster support with supplying a submission script. I attach the `config.yaml` of my currently used profile, that uses the native slurm support, below:

[config.yaml](https://github.com/snakemake/snakemake/files/10353128/config.txt)

**Comment by pvandyken** [comment]
This will be really hard to debug without a good minimal example, especially one that doesn't require external tools. Generating the example is itself a good debugging exercise, as it ensures the problem is indeed with snakemake and not with one of the tools you're using.

**Comment by lhennig** [comment]
Yes, thats true. I'm trying to produce a minimal example at the moment. The main problem is that, even with my scripts above, the weird behavior starts only after 1-2 hours. So creating a minimal example is difficult here, because I always have to wait some hours to see if the minimal example is showing the same weird behavior (and so far, it does not).

But while I'm waiting to see if the minimal example works,  I'm using a debugger to analyze the workflow with my scripts above, and I want to show you the results that I got so far. As far as I understand, there is a Threadpool created in [this](https://github.com/snakemake/snakemake/blob/e03a3b42eea89d512290bf98ee7d77ce2e17447c/snakemake/executors/__init__.py#L500) line that has, in my case, 9 workers (because I specified 4 local cores and there are 5 extra workers created). These workers seem to start localrules in a subprocess in [this line](https://github.com/snakemake/snakemake/blob/e03a3b42eea89d512290bf98ee7d77ce2e17447c/snakemake/executors/__init__.py#L638). I was checking the variable `cmd`, which is given to the subprocess, a couple of times by copying the command and executing it in a new terminal while the snakemake program itself was paused by the debugger. The command always looked like it should just execute the localrule `create_checkpoint_summary`, and this is exactly what it did a lot of times. But yesterday, I was able to get a command from this `cmd` variable that did, indeed, execute `km3sim` locally although it also looked like it should only execute the rule `create_checkpoint_summary`. The command is the following (I used 8 local cores instead of 4 this time, rest should be the same as the scripts above):

`/home/saturn/capn/mppi098h/venvs/woodyNG_3_9_12/bin/python3 -m snakemake --snakefile '/home/saturn/capn/mppi098h/snakemake-workflows/pipelines/neutrino/Snakefile' --target-jobs 'create_checkpoint_summary:dataset=test,p_or_ap=ap,flavor=elec,current=CC,subjob_id=0,e_min=97,e_max=100,dNTrig=20,id=3' --cores 8 --attempt 1 --force-use-threads  --resources 'mem_mb=3000' 'mem_mib=2862' 'disk_mb=1000' 'disk_mib=954' --quiet --force --keep-target-files --keep-remote --max-inventory-time 0 --nocolor --notemp --no-hooks --nolock --ignore-incomplete --rerun-triggers 'mtime' --skip-script-cleanup  --conda-frontend 'mamba' --use-singularity  --wrapper-prefix 'https://github.com/snakemake/snakemake-wrappers/raw/' --local-groupid 'local' --printshellcmds  --latency-wait 86400 --scheduler 'greedy' --scheduler-solver-path '/home/saturn/capn/mppi098h/venvs/woodyNG_3_9_12/bin' --default-resources 'mem_mb=3000' 'disk_mb=max(2*input.size_mb, 1000)' 'tmpdir=system_tmpdir' 'slurm_account=capn' 'slurm_partition=work' --mode 1`

As you can see, it has `create_checkpoint_summary` as target job. I checked whether the files from previous rules for that specific set of wildcards existed, and as it turned out, they did not. I suspect that the command above checks whether the input files for `create_checkpoint_summary` exist, and if they do not, it creates them. If that suspicion is correct, the main question is why snakemake executes `create_checkpoint_summary` although the input files of `create_checkpoint_summary` do not exist.

Below, you can see a screenshot of me entering the command above, which then executes the km3sim rule:

![command_executing_km3sim](https://user-images.githubusercontent.com/95235950/211147613-477f768a-7a1a-4951-a921-4946e5655362.png)

**Comment by pvandyken** [comment]
> As you can see, it has create_checkpoint_summary as target job. I checked whether the files from previous rules for that specific set of wildcards existed, and as it turned out, they did not. I suspect that the command above checks whether the input files for create_checkpoint_summary exist, and if they do not, it creates them. If that suspicion is correct, the main question is why snakemake executes create_checkpoint_summary although the input files of create_checkpoint_summary do not exist.

That sounds like a good start. One thing that may be helpful is to cancel the workflow once you get to the point where you got the above command, then run the command manually. Remove the `--quiet` flags and such, and you should be able to see the DAG and the reasons it wants to run the various outputs.

**Comment by cademirch** [comment]
Hi @lhennig, have you been able to reproduce this with a minimal example? I've run into this recently too, I may have a minimal example too.
